File: 3243-count-the-number-of-powerful-integers/count-the-number-of-powerful-integers.go
func numberOfPowerfulInt(start int64, finish int64, limit int, s string) int64 {
    suffix, _ := strconv.ParseInt(s, 10, 64);

	if finish < suffix {
		return 0;
	}

	lengthSuf := int64(len(s));
	powerOfTen := powInt64(10, int(lengthSuf));
	prefixStart := start / powerOfTen;
	prefixFinish := finish / powerOfTen;

	if start%powerOfTen > suffix {
		prefixStart++;
	}

	if finish%powerOfTen >= suffix {
		prefixFinish++;
	}

	return countValidPrefixes(prefixFinish, limit) - countValidPrefixes(prefixStart, limit);
}

func countValidPrefixes(num int64, limit int) int64 {
	if num == 0 {
		return 0;
	}
	if limit == 9 {
		return num;
	}

	var powLimit [20]int64;
	powLimit[0] = 1;
	for i := 1; i < 20; i++ {
		powLimit[i] = powLimit[i-1] * int64(limit+1);
	}

	var digits []int;
	n := num;
	for n > 0 {
		digits = append(digits, int(n % 10));
		n /= 10;
	}

	for i, j := 0, len(digits)-1; i < j; i, j = i+1, j-1 {
		digits[i], digits[j] = digits[j], digits[i];
	}

	var result int64;
	for i, d := range digits {
		if d > limit {
			return result + powLimit[len(digits) - i];
		}

		result += int64(d) * powLimit[len(digits) - i - 1];
	}

	return result;
}


func powInt64(base, exp int) int64 {
	result := int64(1);

	for i := 0; i < exp; i++ {
		result *= int64(base);
	}

	return result;
}

File: 3243-count-the-number-of-powerful-integers/count-the-number-of-powerful-integers.java
class Solution {
    // Time Limit Exceeded (start = 123456, finish = 32486458654, limit = 4, s = "1")

    public long numberOfPowerfulInt(long start, long finish, int limit, String s) {
        long suffix = Long.parseLong(s);

        if (finish < suffix)
            return 0;

        int lengthSuf = s.length();
        long powerOfTen = (long) Math.pow(10, lengthSuf);

        long prefixStart = start / powerOfTen;
        long prefixFinish = finish / powerOfTen;

        // long result = 0;

        // for (long i = 0; i <= prefixFinish; i++) {
        //     if (isValidPrefix(i, limit)) {
        //         long power = powerOfTen * i + suffix;

        //         if (power > finish) break;

        //         if (power >= start && power <= finish)
        //             result++;
        //     }
        // }

        // return result;

        if (start % powerOfTen > suffix)
            prefixStart++;
        if (finish % powerOfTen >= suffix)
            prefixFinish++;

        return countValidPrefixes(prefixFinish, limit) - countValidPrefixes(prefixStart, limit);
    }

    // private boolean isValidPrefix(long prefix, int limit) {
    //     while (prefix != 0) {
    //         long digit = prefix % 10;
    //         if (digit > limit)
    //             return false;

    //         prefix /= 10;
    //     }

    //     return true;
    // }

    private long countValidPrefixes(long num, int limit) {
        if (num == 0)
            return 0;
        if (limit == 9)
            return num; // tất cả số đều hợp lệ

        int numDigits = (int) Math.log10(num); // số chữ số - 1
        long divisor = (long) Math.pow(10, numDigits);
        long result = 0;

        for (int i = numDigits; i >= 0; i--) {
            int digit = (int) (num / divisor);

            if (digit > limit)
                return result + (long) Math.pow(limit + 1, i + 1); // tất cả tổ hợp còn lại hợp lệ

            result += digit * (long) Math.pow(limit + 1, i); // cộng số tổ hợp nhỏ hơn digit tại vị trí này

            num %= divisor;
            divisor /= 10;
        }

        return result;
    }
}